tests/MongoDB.Driver.Tests/ClientSessionHandleTests.cs

```diff
@@ -258,45 +278,489 @@ public void StartTransaction_should_call_coreSession()
+        private MongoException PrepareException(WithTransactionErrorState state)
+        {
+            var mongoException = new MongoException("test");
+            switch (state)
+            {
+                case WithTransactionErrorState.TransientTransactionError:
+                    {
+                        mongoException.AddErrorLabel(TransientTransactionErrorLabel);
+                        return mongoException;
+                    }
+                case WithTransactionErrorState.UnknownTransactionCommitResult:
+                    {
+                        mongoException.AddErrorLabel(UnknownTransactionCommitResultLabel);
+                        return mongoException;
+                    }
+                case WithTransactionErrorState.NoError:
+                    return null;
+                case WithTransactionErrorState.ErrorWithoutLabel:
+                    {
+                        return mongoException;
+                    }
+            }
+
+            throw new ArgumentException("Not supported ErrorState", state.ToString());
+        }
+
+        private Mock<IClock> CreateClockMock(DateTime now, bool[] isRetryAttemptsWithTimeout, bool shouldNowBeAdded)
+        {
+            if (isRetryAttemptsWithTimeout == null)
+            {
+                isRetryAttemptsWithTimeout = new[] { false };
+            }
+
+            var mockClock = new Mock<IClock>();
+
+            var nowSetup = mockClock.SetupSequence(c => c.UtcNow);
+            if (shouldNowBeAdded)
+            {
+                nowSetup.Returns(now);
+            }
+            foreach (var isTimeoutAttempt in isRetryAttemptsWithTimeout)
+            {
+                var passedTime = CalculateTime(isTimeoutAttempt);
+                nowSetup.Returns(now.AddSeconds(passedTime));
+            }
+            return mockClock;
+        }
+
+        private int CalculateTime(bool timeout)
         {
-            var fieldInfo = typeof(ClientSessionHandle).GetField("_disposed", BindingFlags.NonPublic | BindingFlags.Instance);
-            return (bool)fieldInfo.GetValue(obj);
+            return TransactionExecutorReflector.__transactionTimeout() + (timeout ? 10 : -10);
         }
 
-        public static IServerSession _serverSession(this ClientSessionHandle obj)
+        private void SetupTransactionState(ClientSessionHandle clientSession, bool isTransactionInProgress)
         {
-            var fieldInfo = typeof(ClientSessionHandle).GetField("_serverSession", BindingFlags.NonPublic | BindingFlags.Instance);
-            return (IServerSession)fieldInfo.GetValue(obj);
+            clientSession.WrappedCoreSession.CurrentTransaction.SetState(isTransactionInProgress ? CoreTransactionState.InProgress : CoreTransactionState.Aborted);
         }
     }
-}
+
+    internal static class ClientSessionHandleReflector
+    {
+        public static bool _disposed(this ClientSessionHandle obj) => (bool) Reflector.GetFieldValue(obj, nameof(_disposed));
```

Comment:
Nit: no space between cast and expression being cast.

Reply:
Done